File: 200-Number_of_Islands/main.js
/**
 * 给定一个由 '1'（陆地）和 '0'（水）组成的的二维网格，计算岛屿的数量。一个岛被水包围
 * 并且它是通过水平方向或垂直方向上相邻的陆地连接而成的。你可以假设网格的四个边均被水包围。
 * @param {character[][]} grid
 * @return {number}
 */
var numIslands = function(grid) {
  let island = Array(grid.length);
  
  for (let i = 0; i < grid.length; i++) {
    island[i] = Array(grid[i].length);
  }

  let count = 0;

  for (let i = 0; i < grid.length; i++) {
    for (let j = 0; j < grid[i].length; j++) {
      if (grid[i][j] === '1' && !island[i][j]) {
        count++;
        flagIsland(i,j);
      }
    }
  }

  return count;

  function flagIsland(x, y) {
    if ( x < 0 || x >= grid.length || y < 0 || y >= grid[x].length) {
      return;
    }

    if (grid[x][y] === '1' && !island[x][y]) {
      island[x][y] = true;
      flagIsland(x-1, y);
      flagIsland(x, y+1);
      flagIsland(x+1, y);
      flagIsland(x, y-1);
    }
    return;
  }
};

var arr = [
  ['1', '1', '0', '0', '0'],
  ['1', '1', '0', '0', '0'],
  ['0', '0', '1', '0', '0'],
  ['0', '0', '0', '1', '1']
];

console.log(numIslands(arr));

File: 200-Number_of_Islands/main.go
package main

import (
	"fmt"
)

func numIslands(grid [][]byte) int {
	visited := make([][]bool, len(grid))

	for i := 0; i < len(grid); i++ {
		visited[i] = make([]bool, len(grid[i]))
	}

	count := 0

	for i := 0; i < len(grid); i++ {
		for j := 0; j < len(grid[i]); j++ {
			if grid[i][j] == '1' && visited[i][j] == false {
				count++
				floodfill(visited, grid, i, j)
			}
		}
	}

	return count
}

func floodfill(visited [][]bool, grid [][]byte, x, y int) {
	if x < 0 || x >= len(grid) || y < 0 || y >= len(grid[x]) {
		return
	}

	if grid[x][y] == '1' && visited[x][y] == false {
		visited[x][y] = true
		floodfill(visited, grid, x-1, y)
		floodfill(visited, grid, x, y-1)
		floodfill(visited, grid, x+1, y)
		floodfill(visited, grid, x, y+1)
	}
}

func main() {
	grid := [][]byte{
		[]byte{'1', '1', '0', '0', '0'},
		[]byte{'1', '1', '0', '0', '0'},
		[]byte{'0', '0', '1', '0', '0'},
		[]byte{'0', '0', '0', '1', '1'},
	}

	fmt.Println(numIslands(grid))
}